### ISSUE TITLE
exp: train contacts-and-crops-v1 1.5B on TPU (v5p) following Eric's exp117 best config

### ISSUE THREAD

**Issue by timodonnell** [Open-Athena/MarinFold#137]
## Goal

Train a **1.5B protein language model on the contacts-and-crops-v1 corpus** (exp130/#130, exp132/#132 — the 8k coordinate-document format: contacts + Pass-1 coarse boxes + a few Pass-2 fine crops), **reproducing Eric's exp117 best-so-far training configuration** (MarinFold [#117](https://github.com/Open-Athena/MarinFold/issues/117), marin `eac/plm-exp117`), on **TPU (v5p) on the `marin` iris cluster**.

This is the 8k-crops analog of [#121](https://github.com/Open-Athena/MarinFold/issues/121) (which trains the 32k contacts-and-coordinates-v1 corpus). A convenient property: **crops is an 8192-token format, which is exp117's *native* sequence length** — so Eric's recipe transfers 1:1 with **no RoPE re-tuning** (unlike #121's 32k jump).

---

## Model — Eric's exp117 config, verbatim (do not change the architecture)

Qwen3 **1.47B**, exactly as `MODEL_CONFIG` in `exp117_sweep.py` (marin `eac/plm-exp117`, commit `af2baa4`, 2026-07-15). Same dims as #75 / exp44 / exp67 / exp108 / #121:

```python
Qwen3Config(
    max_seq_len=8192,
    hidden_dim=2048,
    intermediate_dim=8192,
    num_heads=32,
    num_kv_heads=8,        # GQA 4:1, head_dim=64
    num_layers=24,
    rope=Llama3RotaryEmbeddingsConfig(),   # theta 500000
)
```
Qwen3 defaults carry over: `use_qk_norm=True`, RMSNorm, SwiGLU, untied embeddings (~1.47B params). **Only the embedding/output vocab changes** vs contacts-v1: **3848** rows (crops tokenizer) instead of 2845.

**Optimizer / schedule (exp117, unchanged):** AdamW, **cosine** decay, **10% warmup**, `z_loss=None`. Weight decay swept; see best config below. (betas 0.9/0.95, `min_lr_ratio=0.1`, decay excludes norms + embed/output — per #75/#121; confirm exp117's `min_lr_ratio` is levanter default.)

---

## Best configuration from exp117 (current sweep leader)

exp117 is a **live** LR/WD/epochs sweep at fixed Qwen3 1.47B + global batch 128, seq 8192. I queried W&B (`eric-czech/marin`, tag `exp117`, objective `eval/tokenized/contacts-v1-val/loss` at the final step; 222 runs, 138 with a val loss). **Current best *completed* runs:**

| val loss | epochs | LR | WD | global batch | steps | state |
|---:|---:|---|---|---:|---:|---|
| **2.7112** | 16 | **3.1623e-3** | **0.2** | 128 | 71,359 | finished ← leader |
| 2.7131 | 8 | 3.1623e-3 | 0.2 | 64 | 71,359 | finished |
| 2.7267 | 16 | 1e-3 | 1.6 | 128 | 71,359 | finished |
| 2.7267 | 8 | 1e-2 | 0.1 | 128 | 35,679 | finished (8ep/bs128 leader) |

**Recommended anchor: `LR = 3.1623e-3` (√10·1e-3), `WD = 0.2`, global batch 128.** This pair is the robust winner — it leads the 16-epoch rung (2.7112, overall best) *and* the 8-epoch rung (2.7131 at bs64). Caveats to confirm with Eric before launch:
- The sweep is **not finished**: the 32-epoch rung has **0 completed** runs and the 16-epoch rung only **3**. The leader could shift.
- There is real **LR×epochs interaction** (at 8ep/bs128 the winner is `1e-2 / 0.1`, not `3.16e-3 / 0.2`). Since the crops run is long (≥8ep-equivalent, see below), the 16-epoch anchor is the right reference.

---

## Data — contacts-and-crops-v1 (single corpus)

| | value | source |
|---|---|---|
| Bucket root | `hf://buckets/open-athena/MarinFold/data/document_structures/contacts_and_crops_v1/` | exp132 `publish_to_hf.py:28` |
| Splits | `train`, `val`, `test`; `<split>/contacts_and_crops_v1-{shard:05d}-of-{total:05d}.parquet` | `publish_to_hf.py:29,131` |
| Text field | **`document`** | `publish_to_hf.py:144` |
| Docs / tokens | **~4.21M docs, ~34.5B tokens** (~8190 tok/doc; forecast — see repro item 3) | exp132 README |
| Tokenizer | **crops-specific, 3848 vocab** (superset of contacts-v1) | `vocab.py:34`, `core.py:46` |
| Loss | **UNMASKED** (no `<distance>` statements; coordinates are ordinary body tokens) | SPEC; mirror `unmasked_components()` |
| Packing | **`pack=True`, prefix-only, no concat-and-split**; shuffle Feistel `data_seed=0` | `contacts_v1_train_common.py:120-137` |

**Tokenizer detail (matters for warm-start & comparability):** the crops tokenizer is a **strict superset of contacts-v1** — ids **2–2845 are byte-identical** to `timodonnell/contacts-v1-tokenizer`, then ccoord's `<xyz-000..999>` at 2846–3846 (identical ids to #121's tokenizer) and the single new `<crop>` at 3847. Consequence: **contacts-v1 documents tokenize identically**, so we can log `eval/contacts-v1-val/loss` as a **secondary metric directly comparable to Eric's 2.7112** (a superset-softmax caveat, negligible). Recommend tracking both crops-val (primary) and contacts-v1-val (secondary sanity signal).

---

## Token budget / epochs — the one real design change vs exp117

crops is **~34.5B train tokens ≈ 7.4× contacts-v1's 4.68B**, so **Eric's *epoch counts* do not transfer — his *token budget* does.** At batch 128 × seq 8192 = 1.049M tok/step, crops is **~32,900 steps/epoch**.

| match target | tokens | crops epochs | steps |
|---|---:|---:|---:|
| Eric 8ep-equiv | 37.4 B | ~1.1 | ~35,700 |
| **Eric 16ep-equiv (leader) — recommended first run** | **74.8 B** | **~2.15** | **~71,300** |
| #121's 4 epochs | 138 B | 4 | ~131,600 |
| Eric 16 *epochs* literal | ~552 B | 16 | ~526,000 (heavily over-trained; not recommended) |

**Recommendation:** match Eric's 16ep **token budget (~75B, ≈ 2.15 crops epochs)** for the first run — comparable compute to his leader, well past Chinchilla for 1.5B. Cosine decays over that step count. Treat the exact number as an open decision (needs the exact corpus token count, repro item 3).

---

## Compute — TPU v5p on the `marin` iris cluster

Same launch path as exp67/85/108/#121 (marin executor, `default_train`), **not** the exp108 GPU direct-dispatch:

```
uv run iris --cluster marin job run --no-wait --enable-extra-resources \
  --cpu=1 --memory=16GB --disk=16GB --extra=tpu --zone=<region>-a -- python -m <train_module>
```
- **Recommended slice: v5p-128** (≈ 64×H100 bf16); v5p-256 to ~halve wall-clock. At seq 8192 (vs #121's 32768) MFU is higher, so ~75B tokens is a shorter run than #121.
- `ResourceConfig.with_tpu("v5p-128", slice_count=1, ...)` → `SimpleTrainConfig.resources` → `default_train(...)`.
- Framework = Levanter/JAX. Grad-checkpointing on; flash/splash attention.
- Pick the region to co-locate with the crops GCS mirror (repro item 1) and the checkpoint/token-cache prefix.

---

## Reproducibility — prerequisites & gaps (please clear before the full run)

Reproducing Eric's config is fully possible; these are the concrete gaps I found:

1. **Stage the crops corpus to GCS.** It is published **only to the HF bucket**; HF *buckets* are not `HfFileSystem`-addressable for training (same limitation contacts-v1 handles via a byte-identical GCS mirror, `contacts_v1_train_common.py:82-96`). **No crops GCS mirror exists yet** (checked `gs://marin-us-east5/protein-structure/MarinFold/` — exp53/exp105/exp120 present, no crops). Mirror to e.g. `gs://marin-us-east5/protein-structure/MarinFold/exp132_contacts_and_crops_v1/documents/{train,val}/` (region co-located with the TPU zone).
2. **Pin the crops tokenizer.** Unlike contacts-v1 (`timodonnell/contacts-v1-tokenizer@5d68a24a899f`), the crops tokenizer ships only as a `tokenizer/` dir in the corpus bucket — **no HF repo, no pinned revision.** Either publish a pinned tokenizer repo or reference the exact bucket `tokenizer/` path in the trainer (and co-locate it with the model on HF export, per convention).
3. **Compute the exact train-token count.** Only a ~34.5B **forecast** is committed (`stats.json` not checked in, #132 open). Eric hard-codes `TRAIN_TOKENS` to derive steps/epoch precisely; do the same by counting the tokenized crops train cache before fixing `num_train_steps`.
4. **Trainer code path.** Eric's `exp117_sweep.py` uses the **newer** marin API (`marin.experiment.train.train_lm`, `marin.execution.lazy`, `StepRunner`, `fray`). Two faithful routes:
   - **(A, recommended)** reproduce the *config* on MarinFold's proven native path (`marinfold_models.defaults.default_train` + `SimpleTrainConfig`, as exp120/#121) — same recipe, established deps. The full exp117 config is captured in this issue.
   - **(B)** vendor `exp117_sweep.py` + marin's repo-local `experiments/coral/batch_calibration.py` (NOT packaged) and add the `fray` dep, running the newer builder directly — more faithful to code, more moving parts.
5. **Pin marin + commit `uv.lock`** (hard MarinFold rule; the exp108/exp85 marin-drift lesson). Route A needs the marin-core PyPI dev wheels the other model experiments use; Route B needs a marin-core ≥ the 2026-07 build that carries `marin.experiment.train`.
6. **Batch fit.** Route A can hardcode `per_device_parallelism` for the chosen v5p slice from Eric's calibration (v5p family `correction_factor=0.45`; see `exp117_batch_calibration.md`), avoiding the vendored estimator entirely.

---

## Success criteria

1. A converged 1.5B crops model trained with exp117's recipe (`LR 3.16e-3`, `WD 0.2`, batch 128, seq 8192, cosine + 10% warmup, unmasked, `pack=True`, Feistel `data_seed=0`); TPU on `marin`, checkpoints to GCS, W&B (`MarinFold` project).
2. **Two val losses tracked:** `eval/contacts-and-crops-v1-val/loss` (primary) and `eval/contacts-v1-val/loss` (secondary — directly comparable to Eric's **2.7112** thanks to the subset tokenizer).
3. Smoke run first (≤100 steps, real crops data path, cache builds, one checkpoint) before the full run.
4. Report measured s/step, tok/s, MFU, slice size; HF export of the final checkpoint **co-located with the crops tokenizer**.
5. Downstream: run the exp89 pairwise + K=10 contact eval (eval-checkpoint skill) so the crops model is comparable to contacts-v1 models on the 554-protein set.

## Open decisions to confirm before the full run

- **Best config:** confirm `LR 3.1623e-3 / WD 0.2 / bs 128` against Eric's latest sweep state (16/32-epoch rungs still filling in).
- **Token budget:** 16ep-equivalent ~75B (≈2.15 crops epochs, recommended) vs 4 epochs (~138B) vs 8ep-equivalent (~37B)?
- **TPU size:** v5p-128 vs v5p-256; region (drives the GCS mirror target).
- **contacts-v1 mix-in?** Default here is crops-only + contacts-v1 eval-only. Optionally add a token-minority contacts-v1 mix-in (à la #121) to keep pure-contacts capability sharp — decide if wanted.
- **Repro route A vs B** (native `default_train` config-repro vs vendored `exp117_sweep.py`).

## References

- **Eric's config:** MarinFold #117; marin `eac/plm-exp117` @ `af2baa4` — `experiments/protein/{exp117_sweep.py, exp117_policy.md, exp117_batch_calibration.md}`. W&B `eric-czech/marin` (tag `exp117`), leader `prot-exp117-cv1-s02-1_5b-e16-lr3p162e-3-wd0p2-bs128-us-east5` (2.7112). Predecessor #75 winner `prot-exp75-cv1-1_5b-e8-lr1e-3-wd0p2` (2.7566).
- **Dataset:** exp130/#130 (format+SPEC, PR #131), exp132/#132 (corpus). Branch `worktree-exp130-contacts-and-crops`; SPEC `marinfold/marinfold/document_structures/contacts_and_crops_v1/SPEC.md`.
- **Sibling training issue:** #121 (contacts-and-coordinates-v1 32k, from #75).
- **TPU training templates:** exp67 / exp85 (`experiments/exp67_models_contacts_v1_1_5b/`), exp108 (Qwen3 dispatch), exp120 (Qwen3 continue-train on native path). `models/marinfold_models/{defaults.py, simple_train_config.py}`.

---

_Per MarinFold convention, the experiment number = this issue number; use it for the branch, `experiments/exp<N>_models_.../` dir, and run names._

**Comment by timodonnell** [comment]
### Implemented + launching (PR #138)

Built `experiments/exp137_models_contacts_and_crops_v1_1_5b/` — a from-scratch Qwen3 1.47B on contacts-and-crops-v1, reproducing Eric's exp117 leader config (adapted from exp120's native `default_train` path, no warm-start). Config verified locally: LR 3.1623e-3, WD 0.2, bs 128, seq 8192, betas 0.9/0.95, cosine `min_lr_ratio` 0.1, 10% warmup, unmasked, `pack=True`, `data_seed=0`, **num_train_steps = 71,359** (Eric's 16-epoch token budget = ~74.8B tok ≈ 2.21 crops epochs; crops train = 33,824,241,857 tok).

**Reproducibility gaps cleared:**
- **Tokenizer pinned** → `timodonnell/contacts-and-crops-v1-tokenizer@80fe4ee` (3848 vocab, anon-loadable).
- **Corpus staged to GCS** → `gs://marin-us-east5/.../exp132_contacts_and_crops_v1/documents/{train,val,test}/` (2067/22/22 shards, byte-verified vs source). The workstation uplink is ~2.5 MB/s (asymmetric; 55 GB ≈ 6 h), so the mirror ran **on an iris CPU pod** (HF-bucket→GCS, cloud read + cloud write) in ~13 min via `mirror_on_pod.py`.
- **Exact token count committed**; **marin pinned** (`<0.3`), `marinfold-models`@`cc1e478`, `uv.lock` committed.

**Status:** smoke run launched (`exp137-smoke`, v6e-8/us-east5-b, 20 steps). Executor is up and tokenizing the crops corpus (shared cache the full run reuses). Next: confirm from-scratch step-0 loss ≈ ln(3848)≈8.25, then launch the full 71,359-step run on the largest v5p slice that schedules.

Two val losses will be tracked: `eval/contacts-and-crops-v1-val/loss` (primary) and `eval/contacts-v1-val/loss` (secondary, comparable to exp117's 2.7112 via the superset tokenizer).

**Comment by timodonnell** [comment]
### ✅ Full run launched and training

**Smoke run** (`exp137-smoke`, v6e-8) validated the whole pipeline end-to-end: data → tokenize → **from-scratch** train (loss started ~9.4 ≈ ln(3848), i.e. genuine random init, not a warm-start) → eval → checkpoint at step-19 → HF export (tokenizer co-located). The 34B-token crops tokenize cache built in ~7 min across 100+ autoscaled CPU workers and is shared with the full run.

**Full run** — `exp137-cc1-1_5b-lr3p162e-3-wd0p2-bs128` (W&B `open-athena/MarinFold`):
- **Config as spec'd:** from scratch, LR 3.1623e-3, WD 0.2, bs 128, seq 8192, betas 0.9/0.95, cosine `min_lr_ratio` 0.1, 10% warmup, unmasked, `pack=True`, `data_seed=0`, **71,359 steps** (Eric's 16-epoch token budget).
- **Slice:** **v6e-64** @ us-east5-b (co-located with the GCS cache/checkpoints). v5p-128 and v6e-64 in us-east5-a were unallocatable ("no workers match constraints" — no v5p preemptible capacity in-region); the autoscaler booted a v6e-64 in us-east5-b on request. (v5p-32/64 had capacity only in us-central1-a, which would force cross-region cache reads, so I kept it co-located on v6e.)
- **Throughput:** ~**1.5 s/it** steady state → **~30 h** for the full run. Loss descending normally (9.4 → 4.98 by step 120).
- **Checkpoints:** `gs://marin-us-east5/protein-structure/MarinFold/exp137_contacts_and_crops_v1_1_5b/checkpoints/exp137-cc1-1_5b-lr3p162e-3-wd0p2-bs128-95290f/` (permanent every 10k steps + 10-min resume saves; preemption-safe).
- **Two val losses** tracked: `eval/contacts-and-crops-v1-val/loss` (primary) and `eval/contacts-v1-val/loss` (secondary, vs exp117's 2.7112).

Next (after it finishes): HF export of the final checkpoint + downstream exp89 contact eval. PR #138.
